libass/ass_drawing.c

```diff
@@ -43,93 +43,198 @@ static bool token_check_values(ASS_DrawingToken *token, int i, ASS_TokenType typ
     return true;
 }
 
+static inline void add_node(ASS_DrawingToken **tail, ASS_TokenType type, ASS_Vector point)
+{
+    assert(tail && *tail);
+
+    (*tail)->next = calloc(1, sizeof(**tail));
+    if (!(*tail)->next)
+        return;
+    (*tail)->next->prev = *tail;
+    *tail = (*tail)->next;
+    (*tail)->type = type;
+    (*tail)->point = point;
+}
+
+static inline bool get_point(const char **str, ASS_Vector *point)
+{
+    double x, y;
+    if (!mystrtod((char **) str, &x) || !mystrtod((char **) str, &y))
+        return false;
+    *point = (ASS_Vector) {double_to_d6(x), double_to_d6(y)};
+    return true;
+}
+
+
+/**
+ * Parses and advances the string for exactly 3 points.
+ * If the string does not contain fewer than 3 points,
+ * any initial matching coordinates are still consumed.
+ * \return count of added points
+ */
+static bool add_3_points(const char **str, ASS_DrawingToken **tail, ASS_TokenType type)
+{
+    ASS_Vector buf[3];
+
+    if (!*str)
+        return 0;
+
+    bool valid = get_point(str, buf + 0);
+    valid &= get_point(str, buf + 1);
```

Comment:
I want to note that this applies short-circuiting and invalidates the promise given in the function comments that the points are always consumed. It shouldn’t actually affect the overall behaviour though, as these unconsumed points will now be skipped by the main loop if I’m not mistaken. It may be good to adjust the comments to avoid future surprises, though.

(Personally, I think `valid &=` is easier to read than `valid = valid &&`, but I don’t mind either way)

Reply:
> I want to note that this applies short-circuiting and invalidates the promise given in the function comments that the points are always consumed.

wait, is it even possible for any subsequent `get_point` calls to consume anything if a prior one failed? iinm, the first char in the string should then be some non-digit, non-dor and thus any attempt at further processing immediately fail, doesn’t it?

Reply:
> I want to note that this applies short-circuiting and invalidates the promise given in the function comments that the points are always consumed.

To preserve that it's possible to rewrite as
```c
valid = get_point(str, buf + 1) && valid;
```

Reply:
> wait, is it even possible for any subsequent get_point calls to consume anything if a prior one failed?

Ah, no, it is not.